### PR TITLE
Complete #5177

### LINKED ISSUE #4832: Canisters need to be destructible
## Description
when destroyed they should spill gas everywhere all cool like

### PR DESCRIPTION
**Pull request by Vordenburg**
## About the PR 

#5177 only added the DumpCanister behavior for plasma canisters. This PR reduces repetition in the YAML files by handling the spawning of the specific broken canister in the system and moves the DumpCanister behavior to the base gas canister instead.

Resolves #4832, #12877

**Screenshots**

**Changelog**

:cl:
- fix: All gas canisters now dump their contents when destroyed.

### CODE REVIEW

**Review by Mervill — CHANGES_REQUESTED** [review]
Can you chage
```
- type: entity
  parent: BaseStructure
  id: GasCanisterBrokenBase
```
to
```
- type: entity
  parent: BaseStructureDynamic
  id: GasCanisterBrokenBase
```
and add
```
- type: Recyclable
```
to the same entity

File: Content.Server/Atmos/Piping/Unary/Components/GasCanisterComponent.cs [hunk 1]
```diff
@@ -17,6 +19,13 @@ public sealed class GasCanisterComponent : Component, IGasMixtureHolder
         [DataField("container")]
         public string ContainerName { get; set; } = "GasCanisterTankHolder";
 
+        /// <summary>
+        ///     Entity prototype for the broken version of the canister.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("brokenCanister", customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
+        public string BrokenCanister { get; set; } = "GasCanisterBrokenBase";
```
Comment (metalgearsloth) [inline Content.Server/Atmos/Piping/Unary/Components/GasCanisterComponent.cs:27]: Shouldn't destructible be handling this.

File: Content.Server/Atmos/Piping/Unary/Components/GasCanisterComponent.cs [hunk 2]
```diff
@@ -54,7 +63,7 @@ public sealed class GasCanisterComponent : Component, IGasMixtureHolder
         [DataField("releaseValve")]
         public bool ReleaseValve { get; set; } = false;
 
-        [DataField("accessDeniedSound")]
```
Comment (metalgearsloth) [inline Content.Server/Atmos/Piping/Unary/Components/GasCanisterComponent.cs:57]: This is already done implicitly by the type serializer

**Review by metalgearsloth — DISMISSED** [review]
I don't see a reason to move destructible onto entities when it will be more subject to bugs and code quality issues over time.

If you want to reduce repetition in yaml then update the datafields to specify inheritance.

**Comment by Vordenburg** [comment]
> I don't see a reason to move destructible onto entities when it will be more subject to bugs and code quality issues over time.

The Destructible component was already on all of these entities. Is there some better way to handle this behavior of dumping the gas canister's contents and spawning a broken variant than using the Destructible component?

> If you want to reduce repetition in yaml then update the datafields to specify inheritance.

Can you show me what you mean by using the datafields to specify inheritance, such as an example of this being done elsewhere?

The Destructible component was removed from every descendant of GasCanister in order to have a unified specification on the parent, GasCanister, with a field on the GasCanister component itself specifying which broken variant to spawn, instead of having that information needing to be put onto Destructible components on every child with the necessarily duplicated destructible thresholds and et cetera.

**Comment by metalgearsloth** [comment]
> Can you show me what you mean by using the datafields to specify inheritance, such as an example of this being done elsewhere?

https://github.com/space-wizards/RobustToolbox/blob/master/Robust.Shared/Prototypes/IPrototype.cs#L44

> Is there some better way to handle this behavior of dumping the gas canister's contents and spawning a broken variant than using the Destructible component?

No, so destructible should handle it.

**Comment by Ygg01** [comment]
Issue was solved by another PR.

**Comment by Vordenburg** [comment]
> Issue was solved by another PR.

Which PR? If it has been merged, I am still able to break open nitrous oxide canisters and be unaffected.

**Comment by mirrorcult** [comment]
@Ygg01 this was not resolved in another pr

**Comment by EmoGarbage404** [comment]
if the issue was that only the plasma canister had the DumpCanister effect, why not just add it to all of the canisters? I get that it's going to be some yaml bloat, but it's only a few lines and could probably be generalized to all atmos containers, if needed.

**Comment by Vordenburg** [comment]
> if the issue was that only the plasma canister had the DumpCanister effect, why not just add it to all of the canisters? I get that it's going to be some yaml bloat, but it's only a few lines and could probably be generalized to all atmos containers, if needed.

I wasn't aware at the time of writing this PR that this would be the preferred way to solve this problem, so I can change it to that instead. My original submission did generalize the end-result behavior but did not use Destructible.

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**